KEP-2401: Add `TorchTuneConfig` to `train()` API

Electronic-Waste opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow Trainer, check the developer guide:
    https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

This PR adds the initial implementation of `TorchTuneConfig`:

```python
# Configuration for the TorchTune LLM Trainer.
@dataclass
class TorchTuneConfig:
    dtype: Optional[str] = None
    batch_size: Optional[int] = None
    epochs: Optional[int] = None
    loss: Optional[str] = None
    num_nodes: Optional[int] = None
    resources_per_node: Optional[Dict] = None

```

And refactor the `train()` API to add support fine-tuning with `torchtune`. Now, we can specify two types of task in `train()`:

1. **Custom Training Task**: Training with a self-contained function that encapsulates the entire model training process.
2. **Fine-tuning Pre-Trained Models**: Fine-tuning existing pre-trained models.

/cc @kubeflow/wg-training-leads @astefanutti @franciscojavierarceo @deepanker13 @saileshd1402 

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:
Fixes #2504

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/trainer/) included if any changes are user facing

Electronic-Waste committed:
feat(sdk): add TorchTuneConfig.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
rebase(sdk): rebase on the newest master branch.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add args description in train().

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add description for train() func.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): split train() according to trainer and fine_tuning_config

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): update the launching command and args.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add get_args_using_torchtune_config.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): fix some wrong description in train()

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): update the description of fine_tuning_config.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): remove numProcPernode in train().

Signed-off-by: Electronic-Waste <2690692950@qq.com>

andreyvelich commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -190,10 +190,32 @@ def train(
         model_config: Optional[types.HuggingFaceModelInputConfig] = None,
     ) -> str:
-        """Create the TrainJob. TODO (andreyvelich): Add description
+        """
```
Thank you for doing this @Electronic-Waste!
I was thinking a bit more on the API design after our discussion here: https://github.com/kubeflow/trainer/pull/2501#discussion_r1992575796
I feel like we need to be aligned with the trainer_type parameter in the Runtime object.

What do you think about this API design:

```python

@dataclass
class BuiltinTrainer:
    torchtune: Optional[types.TorchTuneConfig]
    nemo: Optional[types.NeMoConfig]
    ...

@dataclass
class CustomTrainer:
    func: Callable = None
    num_nodes: Optional[int] = None
    ...


def train(
    trainer: Optiona[Union[CustomTrainer, BuiltinTrainer]] = None,
    initializer: Optional[Initializer] = None,
    runtime_ref: Optional[str] = None,
) -> str:
    pass
```

So if user wants to use the CustomTrainer, they run:
```python
train(
    trainer=CustomTrainer(
        func=train_func,
        num_nodes=5,
    ),
    initializer=Initializer(
        model=types.HuggingFaceModel(
            storage_uri="hf://gemma-7b",
        ),
    ),
    runtime_ref="torch-distributed",
)
```

If user wants to use Builtin trainer, they run:

```python

train(
    trainer=BuiltinTrainer(
        torchtune=types.TorchTuneConfig(
            dtype="bf32",
            batch_size=128,
            num_nodes=5,
        ),
    ),
    initializer=Initializer(
        dataset=types.HuggingFaceDataset(
            storage_uri="hf://my-dataset",
        ),
    ),
    runtime_ref="torch-tune-llama-7b",
)

```

andreyvelich replied:
Looking for your feedback @kubeflow/wg-training-leads @saileshd1402 @deepanker13 @franciscojavierarceo @astefanutti @akshaychitneni @shravan-achar @seanlaii

Electronic-Waste replied:
> I feel like we need to be aligned with the trainer_type parameter in the Runtime object.

Yeah, I'm also thinking about it. The definition of `trainer` and `fine_tuning_config` has some overlaps, for example, we can also use a custom script for `trainer` to fine-tune the pre-trained model, and fine_tuning_config may not be limited to the builtin trainer in the future. 

The new proposal basically LGTM. I would like to discuss about the API design more:

```python
@dataclass
class BuiltinTrainer:
    post_training: Optional[Union[types.TorchTuneConfig, types.NeMoConfig]]
    ...
```

What about this? Since we may use different trainers in the same framework like HF Trainer and SFTTrainer. Or do you prefer:

```python
@dataclass
class BuiltinTrainer:
    torchtune: Optional[types.TorchTuneConfig]
    nemo: Optional[types.NeMoConfig]
    huggingface:Optional[Union[hf.Trainer, hf.SFTTrainer]]
```

Please let me know your ideas:)

andreyvelich replied:
I like the idea to highlight that these builtin trainers are for post_training, however, maybe it is better to place all of the Builtin trainers under `config`, so it is clear for users that they can select OneOf.

Additionally, we already say in the Runtime object, that it is used for post-training phase.


Should we just do this?
```
@dataclass
class BuiltinTrainer:
    config: Optional[Union[types.TorchTuneConfig, types.NeMoTrainerConfig, types.HFTrainerConfig]]
    ...
```

Electronic-Waste replied:
> Additionally, we already say in the Runtime object, that it is used for post-training phase.

That's true. Most of the trainers are built for post-training. However, there are many training patterns in post-training, for example, SFT, KD, RLHF. And parameters for them varies from each other. Maybe it will be better if we could distinguish them between each others. Like:

```python
@dataclass
class BuiltinTrainer:
    sft: Optional[Union[types.TorchTuneSFTConfig, types.NeMoTrainerConfig]]
    kd: Optional[types.TorchTuneKDConfig]
    rlhf: Optional[Union[types.TorchTuneRLHFConfig, hf.GRPOTrianer]]
    ...
```

It will be better to put them altogether in `config` field. WDYT?

andreyvelich replied:
It would be tricky to support all of them under BuiltinTrainer.

Let's keep the `config` under the `BuiltinTrainer` for now, and if we need to distinguish post-training techniques in the future, we can always update the `TorchTuneConfig` to support it.
I believe the supported algorithms should be driven by BuiltinTrainers' configs (e.g. TorchTuneConfig)

Electronic-Waste replied:
SGTM.

Electronic-Waste committed:
chore(sdk): add TorchTuneConfig in train()

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): add torchtune logic in train()

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
feat(sdk): add BuiltinTrainer.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
feat(sdk): add BuiltinTrainer logic.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add description for initializer.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste commented:
This PR is ready for review now. Looking forward to your feedback! @kubeflow/wg-training-leads @astefanutti

Electronic-Waste committed:
chore(sdk): update description of runtime

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): address unresolved merge error.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste commented:
I've rebased the PR in accordance with the latest changes in master branch. PTAL if you have time!

/assign @kubeflow/wg-training-leads @astefanutti

astefanutti commented on sdk/kubeflow/trainer/constants/constants.py:
```diff
@@ -98,8 +98,61 @@
+
+# The default path to the users' workspace.
+# TODO (andreyvelich): Discuss how to keep this path is sync with pkg.initializers.constants
+WORKSPACE_PATH = "/workspace"
```
It seems some of these environment variables are duplicated like `WORKSPACE_PATH` and may not be in the scope of this PR?

Electronic-Waste replied:
Yeah, that's true. Maybe I need to rebase this PR:)

astefanutti replied:
Thanks, it looks good now!

astefanutti reviewed (commented):
I have a couple of comments about duplicated environment variables and missing LoRA config field, otherwise looks good to me.

astefanutti commented on sdk/kubeflow/trainer/types/types.py:
```diff
@@ -45,10 +45,46 @@ class CustomTrainer:
+    epochs: Optional[int] = None
+    loss: Optional[str] = None
+    num_nodes: Optional[int] = None
+    resources_per_node: Optional[Dict] = None
```
The KEP mentions the `peft_config: LoraConfig` field as well, should it be included there?

Electronic-Waste replied:
Thanks for raising this! We plan to implement lora related logic in: https://github.com/kubeflow/trainer/issues/2505. In current stage, we may want to focus on implementing the whole workflow:)

astefanutti replied:
Ah I missed it, thanks for the clarification.

Electronic-Waste committed:
fix(sdk): remove duplicated fields.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

andreyvelich commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -170,29 +183,63 @@ def train(
+                    trainer.config, types.TorchTuneConfig
+                ):
+                    raise ValueError(
+                        "The config is missing or not supported for the BuiltinTrainer. "
+                        "Currently, we only support `TorchTuneConfig` as parameter."
+                    )
```
Let's make error in a way that, we don't need to always modify it when we support more trainers.
```suggestion
                        "The BuiltinTrainer config is invalid: {trainer.config}"
                    )
```

andreyvelich commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -170,29 +183,63 @@ def train(
+                # Parse args in the TorchTuneConfig to the Trainer, preparing for the mutation of
+                # the torchtune config in the runtime plugin.
+                # Ref:https://github.com/kubeflow/trainer/tree/master/docs/proposals/2401-llm-trainer-v2
+                trainer_crd.command = constants.DEFAULT_TORCHTUNE_COMMAND
+                trainer_crd.args = utils.get_args_using_torchtune_config(trainer)
```
Please can you move the command and args assignment to the helper function, so we will be consistent with
https://github.com/kubeflow/trainer/blob/d5ab621676524130271a121d4292a3ca851b5c3a/sdk/kubeflow/trainer/api/trainer_client.py#L203

```suggestion
                trainer_crd.command, trainer_crd.args = utils.get_args_using_torchtune_config(trainer)
```

Electronic-Waste replied:
Sure!

andreyvelich commented on sdk/kubeflow/trainer/types/types.py:
```diff
@@ -45,10 +45,46 @@ class CustomTrainer:
+        resources_per_node (`Optional[Dict]`): The computing resources to allocate per node.
+    """
+
+    dtype: Optional[str] = None
```
Can you make dtypes, enums similar to: https://github.com/kubeflow/trainer/blob/d5ab621676524130271a121d4292a3ca851b5c3a/sdk/kubeflow/trainer/types/types.py#L95

andreyvelich commented on sdk/kubeflow/trainer/types/types.py:
```diff
@@ -45,10 +45,46 @@ class CustomTrainer:
+    dtype: Optional[str] = None
+    batch_size: Optional[int] = None
+    epochs: Optional[int] = None
+    loss: Optional[str] = None
```
I think, we should make loss enums as well, right ?
https://github.com/pytorch/torchtune/blob/main/recipes/configs/llama3_3/70B_lora.yaml#L58-L59

andreyvelich commented on sdk/kubeflow/trainer/utils/utils.py:
```diff
@@ -310,6 +310,39 @@ def get_entrypoint_using_train_func(
+
+    # Override the loss if it is provided.
+    if fine_tuning_config.loss:
+        args.append(f"loss={fine_tuning_config.loss}")
```
Shoudn't loss be equal to the module name: torchtune.modules.loss.CEWithChunkedOutputLos ?
https://github.com/pytorch/torchtune?tab=readme-ov-file#optimization-flags

Electronic-Waste replied:
Yes, that's true. I'll make loss enums soon.

andreyvelich reviewed (commented):
Thank you for this @Electronic-Waste!
I left my initial thoughts.

Electronic-Waste commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -150,9 +150,22 @@ def train(
         runtime: types.Runtime = types.DEFAULT_RUNTIME,
         initializer: Optional[types.Initializer] = None,
-        trainer: Optional[types.CustomTrainer] = None,
+        trainer: Optional[Union[types.CustomTrainer, types.BuiltinTrainer]] = None,
```
I'll update torch plugin soon: https://github.com/kubeflow/trainer/issues/2507. So I think BuiltinTrainer will work soon  and we can just reserve it.

andreyvelich replied:
@Electronic-Waste Can we implement the logic in the SDK, but not expose the API to the user via `trainer` ?
Given that we suggest users to install SDK from the `master` branch for now, we don't want them to use something which is not ready yet: https://www.kubeflow.org/docs/components/trainer/getting-started/#installing-the-kubeflow-python-sdk

We can just update the `trainer` parameter in the `train()` API in the followup PR once we finish the work.

Electronic-Waste replied:
SGTM. I've updated the API.

Electronic-Waste commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -170,29 +183,63 @@ def train(
+        if trainer:
+            # If users choose to use a custom training function.
+            if isinstance(trainer, types.CustomTrainer):
+                # Add number of nodes to the Trainer.
```
SGTM.

Electronic-Waste commented on sdk/kubeflow/trainer/api/trainer_client.py:
```diff
@@ -170,29 +183,63 @@ def train(
-                    trainer.packages_to_install,
+            # If users choose to use a builtin trainer for post-training.
+            elif isinstance(trainer, types.BuiltinTrainer):
+                if trainer.config is None or not isinstance(
+                    trainer.config, types.TorchTuneConfig
+                ):
```
Get it! I'll wrap them in the new func `get_trainer_from_builtin_trainer()` soon.

Electronic-Waste committed:
fix(sdk): fix train() description according to the review.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add get_trainer_crd_from_custom_trainer

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add get_trainer_crd_from_builtin_trainer and refactor train() API.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
chore(sdk): add Loss enum class.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste commented:
@andreyvelich Thanks for your detailed review! I've addressed all of your comments, except:

- https://github.com/kubeflow/trainer/pull/2522#pullrequestreview-2711850293

We may need more discussions on that.

Electronic-Waste committed:
fix(doc): update loss type in KEP-2401

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): remove BuiltinTrainer in train().

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste commented:
@andreyvelich @astefanutti Thanks for your kind reviews! Please don't hesitate to let me know if you have other suggestions:)

astefanutti commented on sdk/kubeflow/trainer/types/types.py:
```diff
@@ -45,10 +45,54 @@ class CustomTrainer:
+            The number of samples processed before updating model weights.
+        epochs (`Optional[int]`):
+            The number of samples processed before updating model weights.
+        loss (`Optional[str]`): The loss algorithm we use to fine-tune the LLM,
```
Should it be `Optional[Loss]`?

Electronic-Waste replied:
Nice catch!

astefanutti commented on sdk/kubeflow/trainer/types/types.py:
```diff
@@ -45,10 +45,54 @@ class CustomTrainer:
+        resources_per_node (`Optional[Dict]`): The computing resources to allocate per node.
+    """
+
+    dtype: Optional[str] = None
```
Would we also want it to be an enum?

Electronic-Waste replied:
Make sense. I'll update it soon.

Electronic-Waste replied:
I've updated my implementation now. Does it look good to you?👀 @astefanutti

astefanutti replied:
LGTM, just left one final comment to update the proposal accordingly.

Electronic-Waste committed:
fix(sdk): add enum type for dtype.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

astefanutti commented on docs/proposals/2401-llm-trainer-v2/README.md:
```diff
@@ -253,7 +253,7 @@ class TorchTuneConfig:
     dtype: Optional[str] = None
```
```suggestion
    dtype: Optional[DataType] = None
```

Electronic-Waste replied:
Nice catch. I'll also update the doc:)

astefanutti reviewed (approved):
Looks good to me, thanks @Electronic-Waste!

Electronic-Waste committed:
fix(doc): update KEP according to the type update of dtype.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(doc): update the type description in the table.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste committed:
fix(sdk): update dtype validation in utils.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

Electronic-Waste commented:
@astefanutti Thanks for your kind review!

andreyvelich commented on sdk/kubeflow/trainer/utils/utils.py:
```diff
@@ -317,6 +317,101 @@ def get_entrypoint_using_train_func(
+    args = []
+
+    # Override the dtype if it is provided.
+    if fine_tuning_config.dtype:
+        if fine_tuning_config.dtype not in [dtype.value for dtype in types.DataType]:
+            raise ValueError(
+                f"Unsupported dtype: {fine_tuning_config.dtype}. "
+                f"Supported dtypes are: {[dtype.value for dtype in types.DataType]}"
+            )
```
I think, this can be simplified
```suggestion
    if not (fine_tuning_config.dtype, types.DataType):
      raise ValueError(f"Invalid dtype: {fine_tuning_config.dtype})
```

Electronic-Waste committed:
fix(sdk): update dtype override according to the review.

Signed-off-by: Electronic-Waste <2690692950@qq.com>

andreyvelich reviewed (commented):
Thank you for the updates @Electronic-Waste!
/lgtm
/approve

Electronic-Waste commented:
Huge thanks to your patience and detailed reviews! @andreyvelich @astefanutti 

Let's push KEP-2401 forward together!
